66_HLSLBxDFTests/tests.h

```diff
@@ -0,0 +1,704 @@
+#ifndef BXDFTESTS_TESTS_H
+#define BXDFTESTS_TESTS_H
+// cpp only tests
+
+#define GLM_ENABLE_EXPERIMENTAL
+#include <glm/gtx/hash.hpp>
+#include <unordered_map>
+#include <vector>
+#include <cmath>
+#include <format>
+#include <functional>
+
+#include "app_resources/tests_common.hlsl"
+
+// because unordered_map -- next time, do fixed size array of atomic offsets and linked lists (for readback and verification on cpu)
+template<class BxDF, bool aniso = false>
+struct TestBucket : TestBxDF<BxDF>
+{
+    using base_t = TestBxDFBase<BxDF>;
+    using this_t = TestBucket<BxDF, aniso>;
+    using traits_t = bxdf::traits<BxDF>;
+
+    void clearBuckets()
+    {
+        for (float y = -1.0f; y < 1.0f; y += stride)
+        {
+            for (float x = -1.0f; x < 1.0f; x += stride)
+            {
+                buckets[float32_t2(x, y)] = 0;
+            }
+        }
+    }
+
+    float bin(float a)
+    {
+        float diff = std::fmod(a, stride);
+        float b = (a < 0) ? -stride : 0.0f;
+        return a - diff + b;
+    }
+
+    ErrorType compute()
+    {
+        clearBuckets();
+
+        aniso_cache cache;
+        iso_cache isocache;
+
+        sample_t s;
+        quotient_pdf_t pdf;
+        float32_t3 bsdf;
+
+        for (uint32_t i = 0; i < numSamples; i++)
+        {
+            float32_t3 u = ConvertToFloat01<uint32_t3>::__call(base_t::rc.rng_vec<3>());
+
+            if NBL_CONSTEXPR_FUNC (traits_t::type == bxdf::BT_BRDF && !traits_t::IsMicrofacet)
+            {
+                s = base_t::bxdf.generate(base_t::anisointer, u.xy);
+            }
+            if NBL_CONSTEXPR_FUNC (traits_t::type == bxdf::BT_BRDF && traits_t::IsMicrofacet)
+            {
+                if NBL_CONSTEXPR_FUNC (aniso)
+                {
+                    s = base_t::bxdf.generate(base_t::anisointer, u.xy, cache);
+                }
+                else
+                {
+                    s = base_t::bxdf.generate(base_t::isointer, u.xy, isocache);
+                }
+            }
+            if NBL_CONSTEXPR_FUNC (traits_t::type == bxdf::BT_BSDF && !traits_t::IsMicrofacet)
+            {
+                s = base_t::bxdf.generate(base_t::anisointer, u);
+            }
+            if NBL_CONSTEXPR_FUNC (traits_t::type == bxdf::BT_BSDF && traits_t::IsMicrofacet)
+            {
+                if NBL_CONSTEXPR_FUNC (aniso)
+                {
+                    s = base_t::bxdf.generate(base_t::anisointer, u, cache);
+                }
+                else
+                {
+                    s = base_t::bxdf.generate(base_t::isointer, u, isocache);
+                }
+            }
+
+            if (!s.isValid())
+                continue;
```

Comment:
it should be an error to not generate a valid sample if 100% VNDF is guaranteed by the NDF trait

Reply:
https://github.com/Devsh-Graphics-Programming/Nabla-Examples-and-Tests/pull/236#discussion_r2644758827
I guess I'll work on differentiating it in the `sample_type::makeInvalid()` function